ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcIdempotentInsertionsSpec.scala

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.dao
+
+import com.daml.ledger.EventId
+import com.daml.lf.transaction.Node.NodeCreate
+import com.daml.lf.value.Value.ContractId
+import com.daml.platform.ApiOffset
+import org.scalatest.flatspec.AsyncFlatSpec
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.{Inside, LoneElement, OptionValues}
+
+trait JdbcIdempotentInsertionsSpec extends Inside with OptionValues with Matchers with LoneElement {
+  self: AsyncFlatSpec with JdbcLedgerDaoSuite =>
+
+  behavior of "JdbcLedgerDao (on PostgreSQL)"
+
+  it should "allow idempotent transaction insertions" in {
```

Comment:
Nice to see you are including many tests! If you want to improve it further, you could add the following test(s):

1. Test idempotency of partial inserts. E.g., first insert only the state of a transaction (manually simulating a crash before the events were inserted), then insert only the state + events of the same transaction, then insert the whole transaction.
2. Test that the ledger end guards are working. E.g., first insert only the state + events of a transaction (manually simulating a crash before the ledger end was updated), then verify that this transaction can't be looked up.

Reply:
Done. Added the tests in `JdbcPipelinedInsertionsSpec`